include/dlaf/permutations/general/impl.h

```diff
@@ -0,0 +1,202 @@
+    }
+  });
+}
+
+#if defined(DLAF_WITH_CUDA)
+template <class T, Coord coord>
+void applyPermutations(GlobalElementIndex out_begin, GlobalElementSize sz, SizeType in_offset,
+                       const matrix::Distribution& distr, const SizeType* perm_arr,
+                       const std::vector<matrix::Tile<T, Device::GPU>>& in_tiles,
+                       const std::vector<matrix::Tile<T, Device::GPU>>& out_tiles, cudaStream_t stream) {
+  constexpr Coord orth_coord = orthogonal(coord);
+  std::vector<SizeType> splits =
+      util::interleaveSplits(sz.get<orth_coord>(), distr.blockSize().get<orth_coord>(),
+                             distr.distanceToAdjacentTile<orth_coord>(in_offset),
+                             distr.distanceToAdjacentTile<orth_coord>(out_begin.get<orth_coord>()));
+
+  for (SizeType i_perm = 0; i_perm < sz.get<coord>(); ++i_perm) {
+    for (std::size_t i_split = 0; i_split < splits.size() - 1; ++i_split) {
+      SizeType split = splits[i_split];
+
+      GlobalElementIndex i_split_gl_in(split + in_offset, perm_arr[i_perm]);
+      GlobalElementIndex i_split_gl_out(split + out_begin.get<orth_coord>(),
+                                        out_begin.get<coord>() + i_perm);
+      TileElementSize region(splits[i_split + 1] - split, 1);
+      if constexpr (coord == Coord::Row) {
+        region.transpose();
+        i_split_gl_in.transpose();
+        i_split_gl_out.transpose();
+      }
+
+      TileElementIndex i_subtile_in = distr.tileElementIndex(i_split_gl_in);
+      auto& tile_in = in_tiles[to_sizet(distr.globalTileLinearIndex(i_split_gl_in))];
+      TileElementIndex i_subtile_out = distr.tileElementIndex(i_split_gl_out);
+      auto& tile_out = out_tiles[to_sizet(distr.globalTileLinearIndex(i_split_gl_out))];
+
+      dlaf::gpulapack::lacpy<T>(blas::Uplo::General, region.rows(), region.cols(),
+                                tile_in.ptr(i_subtile_in), tile_in.ld(), tile_out.ptr(i_subtile_out),
+                                tile_out.ld(), stream);
+    }
+  };
+}
+#endif
+
+template <Backend B, Device D, class T, Coord C>
+void Permutations<B, D, T, C>::call(SizeType i_begin, SizeType i_end,
+                                    Matrix<const SizeType, Device::CPU>& perms, Matrix<T, D>& mat_in,
+                                    Matrix<T, D>& mat_out) {
+  namespace ut = matrix::util;
+  namespace ex = pika::execution::experimental;
+
+  const matrix::Distribution& distr = mat_in.distribution();
+  TileElementSize sz_last_tile = distr.tileSize(GlobalTileIndex(i_end, i_end));
+  SizeType m = distr.globalTileElementDistance<Coord::Row>(i_begin, i_end) + sz_last_tile.rows();
+  SizeType n = distr.globalTileElementDistance<Coord::Col>(i_begin, i_end) + sz_last_tile.cols();
+  matrix::Distribution subm_distr(LocalElementSize(m, n), distr.blockSize());
+  SizeType ntiles = i_end - i_begin + 1;
+
+  // Note: there is an issue compiling with dlaf::internal::transform()
```

Comment:
Did `transform<false>` not help here or was there another problem?

Reply:
I hadn't merged the teo/permutations branch here. This is a stale comment.